win32ss/gdi/dib/dib32bpp.c

```diff
@@ -269,10 +489,150 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
         }
       }
     }
+    else
+      {
+        /* Buffering for source and destination flip overlaps. Fixes KHMZ MirrorTest CORE-16642 */
+        BOOL OneDone = FALSE;
+
+        if (bLeftToRight)
+        {
+          DPRINT("Flip is bLeftToRight.\n");
+
+          /* Allocate enough pixels for a row in DWORD's */
+          DWORD *store = ExAllocatePoolWithTag(NonPagedPool,
+            (BltInfo->DestRect.right - BltInfo->DestRect.left + 1) * 4, TAG_DIB);
+          if (store == NULL)
+          {
+            DPRINT1("Storage Allocation Failed.\n");
+            return FALSE;
+          }
+
+          /* This sets SourceBits to the bottom line */
+          SourceBits = (PBYTE)BltInfo->SourceSurface->pvScan0
+            + ((BltInfo->SourcePoint.y + BltInfo->DestRect.bottom - BltInfo->DestRect.top - 1)
+            * BltInfo->SourceSurface->lDelta) + 4 * BltInfo->SourcePoint.x;
+
+          /* Set DestBits to bottom line */
+          DestBits = (PBYTE)BltInfo->DestSurface->pvScan0
+           + (BltInfo->DestRect.bottom - 1) * BltInfo->DestSurface->lDelta
+           + 4 * BltInfo->DestRect.left;
+
+          for (j = BltInfo->DestRect.bottom - 1; BltInfo->DestRect.top <= j; j--)
+          {
+
+              /* Set Dest32 to right pixel */
+              Dest32 = (DWORD *) DestBits + (BltInfo->DestRect.right - BltInfo->DestRect.left - 1);
+              Source32 = (DWORD *) SourceBits;
+
+              Index = 0;
+
+              /* Store pixels from left to right */
+              for (i = BltInfo->DestRect.right - 1; BltInfo->DestRect.left <= i; i--)
+              {
+                store[Index] = XLATEOBJ_iXlate(BltInfo->XlateSourceToDest, *Source32++);
+                Index++;
+              }
+
+              Index = 0;
+
+              /* Copy stored dat to pixels from right to left */
+              for (i = BltInfo->DestRect.right - 1; BltInfo->DestRect.left <= i; i--)
+              {
+                *Dest32-- = XLATEOBJ_iXlate(BltInfo->XlateSourceToDest, store[Index]);
+                Index++;
+              }
+            SourceBits -= BltInfo->SourceSurface->lDelta;
+            DestBits -= BltInfo->DestSurface->lDelta;
+          }
+          ExFreePoolWithTag(store, TAG_DIB);
+          OneDone = TRUE;
+        }
+
+        if (bTopToBottom)
+        {
+
+          /* Note: It is very important that this code remain optimized for time used. */
+          /*   Otherwise you will have random crashes in ReactOS that are undesirable. */
+          /*   For an example of this just try executing the code here two times.      */
```

Comment:
```suggestion
          /* Note: It is very important that this code remain optimized for time used.
           *   Otherwise you will have random crashes in ReactOS that are undesirable.
           *   For an example of this just try executing the code here two times. */
```